Move `RestartDelayedEventServlet` to workers

anoadragon453 opened the pull request:
To allow delayed event timeouts to be extended even if the main process is currently restarting (for instance, during a pod resize). Addresses the `/restart` portion of https://github.com/element-hq/synapse/issues/19141.

Note that this *does not* move the older, general `POST /org.matrix.msc4140/delayed_events/{delayID}` endpoint to workers. That endpoint had the caller specify `{restart,send,cancel}` as a JSON body parameter. As such, clients will need to ensure they're using `/restart` to get the benefit of worker request handling.

Part of MSC4140

### Pull Request Checklist

<!-- Please read https://element-hq.github.io/synapse/latest/development/contributing_guide.html before submitting your pull request -->

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

anoadragon453 committed:
Move `RestartDelayedEventServlet` to workers

To allow delayed event timeouts to be extended even if the main process
is currently restarting (for instance, during a pod resize).

anoadragon453 committed:
newsfile

anoadragon453 committed:
Add entry to upgrade notes

devonh commented on synapse/handlers/delayed_events.py:
```diff
@@ -81,6 +82,15 @@ def __init__(self, hs: "HomeServer"):
+        # while the main process is pulling delayed events from the DB, the two
+        # will not race, as the main process will mark the delayed events as
+        # processed in the DB immediately before sending them.
+        self._initialized_from_db = defer.succeed(None)
```
I think the changes here are okay, but I'm mostly taking you at your word that this comment is true.

Would you be able to provide a more thorough explanation / links to the relevant code which validates this?

anoadragon453 replied:
Ah sorry. See the code just below where `self._initialized_from_db` is set. I failed to realise that someone reading this from top-to-bottom wouldn't yet know what `self._initialized_from_db` was.

Essentially `_initialized_from_db` is a deferred that completes once all pending delayed events in the DB have been scheduled. The main process will manage that. The `restart` handler will update the timeout in the DB, so is not dependent 

It's probably simpler if `await make_deferred_yieldable(self._initialized_from_db)` below is only called on the main process. I've made that visual change in 6f6227c40b.

We're effectively letting the `restart` handler code on workers and `_schedule_db_events` on the main process run in parallel. I had a think about whether this would result in a race condition.

The `restart` handler calls [`restart_delayed_event`](https://github.com/element-hq/synapse/blob/6f6227c40b5bf6b3ec10d874572a04ca6acff429/synapse/storage/databases/main/delayed_events.py#L164-L208), which updates a delayed event's timeout as long as it hasn't been marked as processed (`is_processed`) yet.

`is_processed` means "sending of this event has begun". Once a delayed event has been successfully sent, it is deleted from the `delayed_events` table.

I don't believe there's a point between the queries in `_schedule_db_events` where running `UPDATE delayed_events SET send_ts = ? + delay WHERE delay_id = ? AND NOT is_processed` would be detrimental.

The only thing that stands out to me is between `self._store.unprocess_delayed_events` and `self._store.process_timeout_delayed_events`, when all events marked as "processed" in the DB are set to unprocessed again (because processing may have failed partway during homeserver shutdown). But in that case, the event in question will just have its `send_ts` successfully extended and `self._store.process_timeout_delayed_events` won't send it yet, which sounds like what the client intended anyhow.

And then given that, if the `restart code can't race with the initialisation code anyhow, we could just remove `await make_deferred_yieldable(self._initialized_from_db)` from `restart` altogether, even when running on the main process.

cc @AndrewFerr who originally wrote this code.

benbz commented on docs/workers.md:
```diff
@@ -285,6 +285,9 @@ information.
     ^/_matrix/client/(r0|v3|unstable)/user_directory/search$
 
+    # Unstable MSC4140 support
+    ^/_matrix/client/unstable/org.matrix.msc4140/delayed_events/.*/restart$
```
```suggestion
    ^/_matrix/client/unstable/org.matrix.msc4140/delayed_events(/.*/restart)?$
```

I was PR a separate change to:
* Remove line 294 below
* Add `^/_matrix/client/unstable/org.matrix.msc4140/delayed_events$` up in this block

As currently the single worker-able endpoint is also the only one without the delay ID and it is clearer to have it up here rather than in the `GET` only block below.

Given my proposed PR would conflict with this, making the suggestion here

anoadragon453 committed:
Move delayed_events GET to "MSC4140" section

GET is the only method on `/delayed_events$`, so it's fine to put it
into the above section.

Thanks Ben BZ for the suggestion.

anoadragon453 committed:
Modify Complement config

So that we test the GET `/delayed_events$` endpoint on workers as well.

anoadragon453 committed:
Rework await-ing initialisation logic to be clearer

Rework the code to be clearer. There should be no functional difference in this commit.

anoadragon453 committed:
Only schedule delayed events on the main process

The main process won't miss sending the event. When the original timeout
is hit, the main process will simply pick up the new timeout and
schedule a timer for that moment instead.

anoadragon453 committed:
Add `reprocess_events` to `process_timeout_delayed_events`

per @AndrewFerr's suggestion in https://github.com/element-hq/synapse/pull/19207#discussion_r2556631093

anoadragon453 committed:
Drop `_initialized_from_db` await in `restart`

anoadragon453 committed:
re-add note regarding double-sending events

anoadragon453 commented on synapse/handlers/delayed_events.py:
```diff
@@ -96,16 +96,13 @@ async def _schedule_db_events() -> None:
-                # Caveat: this will double-send delayed events that successfully persisted, but failed
-                # to be removed from the DB table of delayed events.
-                # TODO: To avoid double-sending, scan the timeline to find which of these events were
-                # already sent. To do so, must store delay_ids in sent events to retrieve them later.
```
Ah, yes it will. Good catch. Re-added!

anoadragon453 committed:
Fix syntax typo in SQL logic

devonh reviewed (approved):
Awesome, this looks a lot better now with easier to understand rationale for why we can rely on certain behaviour. :)

anoadragon453 committed:
Merge branch 'develop' of github.com:element-hq/synapse into anoa/delayed_events_restart_workers
